Change where extension IPC is exported

Nokel81 opened the pull request:
- Now under Ipc.IpcMain and Ipc.IpcRenderer

- Allow listener's to be cleaned up early by extensions

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Change where extension IPC is exported

- Now under Ipc.IpcMain and Ipc.IpcRenderer

- Allow listener's to be cleaned up early by extensions

Signed-off-by: Sebastian Malton <sebastian@malton.name>

stefcameron commented on src/extensions/ipc/ipc-main.ts:
```diff
@@ -19,27 +19,43 @@
+   * @param channel The name of the RPC
+   * @param handler The remote procedure that is called
+   */
+  handleIpc(channel: string, handler: (event: Electron.IpcMainInvokeEvent, ...args: any[]) => any): void {
```
Why aren't we returning a disposer in this case too?

Nokel81 replied:
I can add some documentation here. But RPC is quite different than IPC. I think I will change the method names too.

Namely, because it is an error to call `invokeRpc` without an "alive" `handleRpc` it doesn't lend itself to being as transient as `listenIpc`. 

Furthermore, the use case surrounding early cleanup of `listenIpc` doesn't apply to `handleRpc` as `main` has no concept of components mounting and unmounting on the DOM.

stefcameron replied:
OK, makes sense. Some mention in the docs about this would good.

stefcameron commented on src/extensions/ipc/ipc-renderer.ts:
```diff
@@ -19,23 +19,40 @@
+   * @param args The arguments to pass to the RPC
+   * @returns A promise of the resulting value
+   */
   invokeIpc(channel: string, ...args: any[]): Promise<any> {
```
Do we call `invokeIpc()` regardless of how the channel was declared in `main` (i.e. same method whether it's a listener or a handler)? Probably not since there's still a `broadcastIpc()` in the `IpcRegistrar`, I suppose.

Nokel81 replied:
No, `invokeIpc` (I will rename this to `invokeRpc`) *only* interacts with `handleRpc` (formerly `handleIpc`).

stefcameron reviewed (commented):
The listeners feel a lot like event listeners, so it feels like there should be a pair of methods to add and remove listeners, instead of the add returning a disposer that needs to be invoked (and therefore needs to be kept in scope if a component is adding a listener on mount and then removing it on unmount). Disposer is OK, but doesn't feel natural to me.

I think this PR should also be updating the docs here: https://github.com/lensapp/lens/pull/2775/files#diff-35cfa0896b5dd8347734ddd548f83c3869571d486e0651c7c0e7d016b5194bd2R34

Things have changed.

And we should take this opportunity to remove the unnecessary `registerListeners` import from this line too: https://github.com/lensapp/lens/pull/2775/files#diff-35cfa0896b5dd8347734ddd548f83c3869571d486e0651c7c0e7d016b5194bd2R46

Nokel81 committed:
add some more docs, change names to handleRpc and invokeRpc

Signed-off-by: Sebastian Malton <sebastian@malton.name>

stefcameron commented on docs/extensions/guides/ipc.md:
```diff
@@ -112,10 +112,10 @@ As this example shows: the channel names *must* be the same.
 The cleanup of these handlers is handled by Lens itself.
 
-`Store.RendererIpcStore.broadcastIpc(...)` and `Store.MainIpcStore.broadcastIpc(...)` sends an event to all renderer frames and to main.
+`Ipc.IpcRenderer.broadcastIpc(...)` and `Ipc.IpcMain.broadcastIpc(...)` sends an event to all renderer frames and to main.
```
I would say, "sends an event" -> "send events", since we have 2 things we're talking about.

stefcameron replied:
And then, couldn't these calls be `IcpRenderer.getInstance().broadcastIcp(...)` and `IcpMain.getInstance().broadcastIpc()`?

stefcameron commented on docs/extensions/guides/ipc.md:
```diff
@@ -128,4 +128,4 @@ This means that more types than what are JSON serializable can be used, but not
 Calling IPC is very simple.
 If you are meaning to do an event based call, merely call `broadcastIpc(<channel>, ...<args>)` from within your extension.
```
"`broadcastIpc(<channel>, ...<args>)`" -> "`Ipc.IcpMain.broadcastIpc(<channel>, ...<args>)` (`IpcRenderer` could also be used here, as there would be no difference in the result)"

stefcameron reviewed (commented):
- [This title](https://github.com/lensapp/lens/pull/2845/files/ccac859b0ba585bde42f89735b77c01c37c70bd2..fa87d9edbdb109b194e2b0b926a6e9be48d2c04f#diff-35cfa0896b5dd8347734ddd548f83c3869571d486e0651c7c0e7d016b5194bd2R25) should be renamed "Request Based RPC" now.
- Should [this title](https://github.com/lensapp/lens/pull/2845/files/ccac859b0ba585bde42f89735b77c01c37c70bd2..fa87d9edbdb109b194e2b0b926a6e9be48d2c04f#diff-35cfa0896b5dd8347734ddd548f83c3869571d486e0651c7c0e7d016b5194bd2R126) now be "Using IPC and RPC"?

I'm not sure renaming `handleIpc` to `handleRpc` made things better. Technically, it's correct, I guess, since RPC is "Remote Procedure Call" and you would expect a response from calling that procedure.

But it has created this IPC/RPC jumble, like `Ipc.IpcRenderer.handleRpc()` -- wait, what?

Does the extension need to really understand the deep difference between the two? If I wanted to, I could use a handler to send a message, and just not `await` it, knowing that this particular channel handler just responds with `undefined` anyway (same as calling any function in JavaScript that doesn't explicitly return a value; all functions technically return a value even if they don't have a `return` statement in them). It would be no different from broadcasting to a single listener on a given channel.

What if we dropped the `Ipc` and `Rpc` postfixes (since we're renaming things anyway)? You either `listen()` or `handle()`, and then either `broadcast()` or `invoke()`. I don't think, as an extension developer, I really need to care about whether I'm doing IPC or RPC, do I? The gist of it is that I'm making some type of cross-process remote call. In either case, the same object serialization algorithm is at play for any parameters I pass. The only differences I see are:
- `broadcast()` sends to _all_ listeners of the same channel and never responds (so no point `await`-ing it)
- `invoke()` probably sends to the handler and responds (so I should probably `await` it)

`class IpcRenderer extends Ipc.IpcRenderer` -> could we have `Ipc.Renderer` here, dropping the redundant `Ipc`? Maybe it's the name of the class in the extension that's making this a bit confusing, because in the docs, we end-up with two distinct `IpcRenderer` classes to talk about. If not `Ipc.Renderer` (and `Ipc.Main`), then how about `class RemoteRenderer extends Ipc.Ipc.Renderer`?

And then it's still strange to me to do `RemoteRenderer.getInstance().invokeRpc()` (or even if it were `RemoteRenderer.getInstance().invoke()`). It feels like I'm invoking on the `renderer`, when I'm actually invoking on `main`. I want to do `RemoteMain.getInstance().invoke()` -- from my code running in `renderer` (because I'm invoking on the `main` thread, not on the `renderer` thread, because I know the handler is declared on `main`). Maybe that's already possible? Not sure.

Nokel81 committed:
Review comment changes

- Remove Rpc and Ipc suffixes from methods

- Remove Ipc prefix from exported class names

- Fixed up docs to be correct

Signed-off-by: Sebastian Malton <sebastian@malton.name>

stefcameron commented on docs/extensions/guides/ipc.md:
```diff
@@ -112,20 +112,17 @@ As this example shows: the channel names *must* be the same.
 The cleanup of these handlers is handled by Lens itself.
 
-`Ipc.IpcRenderer.broadcastIpc(...)` and `Ipc.IpcMain.broadcastIpc(...)` sends an event to all renderer frames and to main.
+Calling either `IpcRenderer.getInstance().broadcast(...)` or `IpcMain.getInstance().broadcast(...)` sends an event to all renderer frames and to main.
```
Since we used backticks on `main` everywhere else, the one reference here at the end of the sentence should have them too.

stefcameron reviewed (commented):
- Should [this title](https://github.com/lensapp/lens/pull/2845/files#diff-35cfa0896b5dd8347734ddd548f83c3869571d486e0651c7c0e7d016b5194bd2R25) be "Request Based RPC"?
- I think [this section](https://github.com/lensapp/lens/pull/2845/files#diff-35cfa0896b5dd8347734ddd548f83c3869571d486e0651c7c0e7d016b5194bd2R34) is still not updated for this IPC vs RPC distinction.
- We're missing mention of the disposer when declaring a listener, though I suppose that will come out in the docs based on the typings. Just mentioning in case you wanted to point this out now, since this PR also adds that feature.

Nokel81 committed:
some more doc fixes

Signed-off-by: Sebastian Malton <sebastian@malton.name>

stefcameron commented on docs/extensions/guides/ipc.md:
```diff
@@ -14,22 +15,22 @@ There are two flavours of communication that are provided:
-This is accomplished by returning a `Promise<T>` which needs to be `await`-ed.
+This is more like a Remote Procedure Call (RPC) or Send-Receive-Reply (SRR).
+With this sort of communication the caller needs to wait for the result from the other side.
+This is accomplished by `await`-int the returned `Promise<any>`.
```
"int" -> "ing"

stefcameron commented on docs/extensions/guides/ipc.md:
```diff
@@ -14,22 +15,22 @@ There are two flavours of communication that are provided:
 
 This is a unidirectional form of communication.
-Only `LensRendererExtension` can initiate this kind of request, and only `LensMainExtension` can and respond this this kind of request.
+Only `renderer` can initiate this kind of request, and only `main` can and respond this this kind of request.
```
"respond this" -> "respond to this"

stefcameron commented on docs/extensions/guides/ipc.md:
```diff
@@ -75,7 +76,7 @@ function onInitialize(event: Types.IpcMainEvent, id: string) {
 
 In other files, it is not necessary to pass around any instances.
-It should be able to just call `getInstance()` everywhere in your extension as needed.
+It should be able to just call `IpcMain.getInstance()` anywhere it is needed in your extension.
```
"It should be able" -> "You should be able"?

stefcameron reviewed (commented):
✅ I found a few more typos, but other than that, this is looking solid now! Thanks for working through all this with me. Hopefully the rest of the team likes these changes too.

Nokel81 committed:
more doc fixes

Signed-off-by: Sebastian Malton <sebastian@malton.name>

jakolehm reviewed (approved):
LGTM

jim-docker commented on docs/extensions/guides/ipc.md:
```diff
@@ -1,35 +1,36 @@
 
 This is a unidirectional form of communication.
-Only `LensRendererExtension` can initiate this kind of request, and only `LensMainExtension` can and respond this this kind of request.
+Only `renderer` can initiate this kind of request, and only `main` can and respond to this kind of request.
```
```suggestion
Only `renderer` can initiate this kind of request, and only `main` can receive and respond to this kind of request.
```

jim-docker replied:
```suggestion
Only `renderer` can initiate this kind of request, and only `main` can receive and respond to this kind of request.
```

jim-docker commented on src/extensions/ipc/ipc-main.ts:
```diff
@@ -19,27 +19,43 @@
+   */
+  listen(channel: string, listener: (event: Electron.IpcRendererEvent, ...args: any[]) => any): Disposer {
     const prefixedChannel = `extensions@${this[IpcPrefix]}:${channel}`;
+    const cleanup = once(() => ipcMain.removeListener(prefixedChannel, listener));
```
nice

jim-docker commented on src/extensions/ipc/ipc-renderer.ts:
```diff
@@ -0,0 +1,65 @@
+   * a component then putting the returned value in a `disposeOnUnmount` call will suffice.
+   * @param channel The channel to listen for broadcasts on
+   * @param listener The function that will be called with the arguments of the broadcast
+   * @returns An optional disopser, Lens will cleanup even if this is not called
```
```suggestion
   * @returns An optional disposer, Lens will cleanup even if this is not called
```
